# Week 3 — Data Preprocessing & Outlier Detection

## Objective
Assess missing values, duplicates, datatype issues and statistical outliers, then document the cleaning strategy.

In [1]:
matches.isna().sum().sort_values(ascending=False)

Matches missing values:
umpire3            637
city                 7
winner               4
player_of_match      4
umpire1              2
umpire2              2


In [1]:
df.duplicated().sum()

Duplicate rows in all four datasets: 0


In [1]:
matches['date'] = pd.to_datetime(matches['date'], errors='coerce')
batsmen['HS_numeric'] = pd.to_numeric(
    batsmen['HS'].astype(str).str.replace('*','',regex=False),
    errors='coerce'
)

Datatype standardisation completed: matches.date → datetime; batsmen.HS_numeric → numeric.


In [1]:
iqr_outliers(batsmen, batsmen.select_dtypes(include='number').columns)

Numerical outlier counts for batsmen were calculated using the 1.5×IQR rule.


In [1]:
iqr_outliers(bowlers, bowlers.select_dtypes(include='number').columns)

Numerical outlier counts for bowlers were calculated using the 1.5×IQR rule.


## Cleaning Strategy

- `umpire3`: 637 missing values. Treat as not recorded; use `Unknown` if retained, or exclude it from player-valuation modelling if irrelevant.
- `city`: 7 missing values. Use `Unknown` if city analysis is required.
- `winner` and `player_of_match`: 4 missing values each. Do not blindly impute.
- `umpire1` and `umpire2`: 2 missing values each. Use `Unknown` if retained.
- Convert `date` to datetime.
- Convert batsmen `HS` to a numeric analysis field after removing `*`.
- Standardise player names before merging datasets.
- Review IPL team-name changes across seasons before joining datasets.
- Flag outliers rather than automatically deleting them because exceptional IPL performances can be genuine.